### PR TITLE
Add Open Addressing Hash Map

### LINKED ISSUE #11: Implement HashTable with Open Addressing

### PR DESCRIPTION
**Pull request by hungcq**

### CODE REVIEW

**Review by cuonghq104 — APPROVED** [review]
Great!!

**Comment by trungducc** [comment]
@vietduc01100001 Please help to take a look

File: src/Array.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
+/**
+ * Created by: HungCQ
+ * Date: 20-May-20
+ */
+public class Array {
+    private int size;
+    private int[] arr;
+
+    public Array() {
+        arr = new int[0];
```
Comment (hellovietduc) [inline src/Array.java]: Initializing the internal array to be of length 0 is a bad practice. It means for the first 3 items inserted into the array, we have to grow the capacity 3 times! Choose a power of 2 number, 8 or 16 would be a good choice.

File: src/Array.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    public int capacity() {
+        return arr.length;
+    }
+
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    public int itemAt(int index) {
+        if (index > size - 1) {
```
Comment (hellovietduc) [inline src/Array.java]: What if we pass -1, -2 or -3000 to this function? Would it throw an error? Or should we handle it nicely?

File: src/Array.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    public void append(int item) {
+        size++;
+        checkAndIncreaseSize();
+        arr[size - 1] = item;
+    }
+
+    public boolean insert(int item, int index) {
+        if (index == size) {
```
Comment (hellovietduc) [inline src/Array.java]: Same as above, you should validate the `index` parameter.

File: src/Array.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        }
+        arr[index] = item;
+        return true;
+    }
+
+    private void checkAndIncreaseSize() {
+        if (arr.length == 0) {
+            arr = new int[1];
+            return;
+        }
```
Comment (hellovietduc) [inline src/Array.java]: You won't need to handle this edge case if you choose a good default capacity.

File: src/Array.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        if (size > arr.length) {
+            int[] newArr = new int[arr.length * 2];
```
Comment (hellovietduc) [inline src/Array.java]: Avoid hard-coding the grow factor `2` into the function.

File: src/Array.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    public int pop() {
+        int value = arr[size - 1];
+        arr[size - 1] = 0;
+        size--;
+        checkAndReduceSize();
+        return value;
+    }
+
+    public int removeAt(int index) {
+        if (index >= size - 1) {
```
Comment (hellovietduc) [inline src/Array.java]: Validate the `index` parameter when it's a negative integer.

File: src/Array.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        }
+        arr[size - 1] = 0;
+        size--;
+        checkAndReduceSize();
+        return value;
+    }
+
+    private void checkAndReduceSize() {
+        if (size <= arr.length / 4) {
+            int[] newArr = new int[arr.length / 2];
```
Comment (hellovietduc) [inline src/Array.java]: Same as above, avoid hard-coding "magic numbers".

**Review by hellovietduc — CHANGES_REQUESTED** [review]
Please create separate PRs for each issue. Don't mix Array or Linked List in Hash Map PR.

Plus, please add test cases and if possible set up GitHub Actions to run it automatically, and everyone can see the result as well.

**Pull request closed without merging**